tests/test_responsetypes.py

```diff
@@ -120,8 +120,15 @@ def test_from_args_post_xtractmime(self):
                                                        'Content-Type': ['text/html']})}, HtmlResponse),
             ({'body': b'Some plain text', 'headers': Headers({'Content-Type': 'application/octet-stream'})},
              TextResponse),
-            ({'url': b'http://www.example.com/page/file.html', 'headers': Headers({'Content-Type': 'application/octet-stream'})},
-             TextResponse),
+            ({'url': b'http://www.example.com/page/file.html',
+              'headers': Headers({'Content-Type': 'application/octet-stream'})}, TextResponse),
+            ({'url': 'http://www.example.com/item/file.xml',
+              'headers': Headers({'Content-Type': 'application/octet-stream'})}, TextResponse),
+            ({'url': 'http://www.example.com/item/file.html',
+              'headers': Headers({'Content-Type': 'application/octet-stream'})}, TextResponse),
+            ({'url': 'http://www.example.com/item/file.xml',
+              'headers': Headers({'Content-Disposition': ['attachment; filename="data.xml.gz"'],
+                                  'Content-Type': 'application/octet-stream'})}, TextResponse),
```

Comment:
I think this change is something we should address, in the policy of not going backwards when it comes to finding a more specific response class.

Reply:
> I think this change is something we should address, in the policy of not going backwards when it comes to finding a more specific response class.

By this you mean to change ` _guess_content_type`, right? or I am interpreting something wrong?

Reply:
I was not pointing at any specific way to address it, but yes, I think that’s where the changes need to be made.

Reply:
Hey @Gallaecio, I was reviewing the `_guess_content_type` function. To resolve the above test issue, should I prefer (filename or URL) over the `Content-Type` header in the `if` statements? Or Please suggest a way to resolve the issue if you have any changes in mind.

Reply:
Oh my, it took me a while to remember what this was about :facepalm: 

OK, so I understand that pre-xtractmime this test case would result in `XmlResponse`, but now it causes `TextResponse`. And we do not want that, because if before a more specific class was selected and made sense, going to a more generic class now (`XmlResponse` → `TextResponse`) may be considered a regression from the user point of view.

Looking at `_guess_content_type`, I think we may need to refactor it to run the following logic instead when determining the value of the content-type to pass later to xtractmime:

```python
content_type_mime_type = ...
content_disposition_mime_type = ...
url_mime_type = ...

prioritized_mime_type_checkers = (
    is_html_mime_type,
    is_xml_mime_type,
    _is_text_mime_type,
)
candidate_mime_types = (
    content_type_mime_type,
    content_disposition_mime_type,
    url_mime_type,
)
for mime_type_checker in prioritized_mime_type_checkers:
    for candidate_mime_type in candidate_mime_types:
        if (
            candidate_mime_type is not None 
            and mime_type_checker(candidate_mime_type)
        ):
            return candidate_mime_type
return (
    content_type_mime_type
    or content_disposition_mime_type
    or url_mime_type
)
```

This is untested, incomplete code, of course. `prioritized_mime_type_checkers` can be defined outside this function, its 2 first values are from xtractmime, and `_is_text_mime_type` can be refactored out of `_guess_response_type` (the `if` that returns `TextResponse`).

But I hope it gets the general idea through:
1. Get all possible MIME types first
2. Choose the one matching the most specific response class: HTML, XML, text, binary.

Reply:
Yeah, It took me a while too to remember everything 😅 . 
Anyways, thank you so much for this. I will test this and let you know if things are working.